### PR TITLE
Receiving 401 responses on dyno shutdown

### LINKED ISSUE #58: Ignore missing dyno
## Expected

When the app is properly configured and Heroku is working properly barnes should not warn

## Actual

A customer reported that they're using preboot and that every time their dyno restarts they get this warning:

```
barnes: metrics POST rejected (401): dynoidmap: no dyno information exists for that UUID
```

## Discussion

This is possibly a race condition where the old dyno is removed from the registry before it is fully stopped

This can ALSO occur when there is a platform incident, and restarting a dyno (once the incident is resolved) can fix the issue. However, this is such a rare occurrence (internal note: search "401 dynoidmap" in support cases) that I don't think it's worth suggesting people try to restart their dyno.

Internal links:

- https://heroku.support/1696566

## Suggested

Cool down the warning such that it isn't emitted right away (if a 401 with this message is expected on dyno shutdown). Preserve emitting the warning on BARNES_DEBUG or after N occurances of a 401.

### PR DESCRIPTION
**Pull request by schneems**
When the dyno is shutting down, there's a race condition where the component might be unregistered with the API that Heroku uses to capture metrics before the dyno finishes fully shutting down. That means there's a window where a dyno still going through a clean SIGTERM shutdown could still be sending metrics, but the API doesn't recognize the dyno.

To avoid this from cluttering logs when the dyno is restarted, delay printing 401 warnings for 90 seconds. This will give the dyno time to shut down safely with some buffer room https://devcenter.heroku.com/articles/error-codes#r12-exit-timeout. Anyone debugging connection issues via BARNES_DEBUG will see the warnings right away.

Closes #58

### CODE REVIEW

File: barnes.gemspec [hunk 1]
```diff
@@ -32,4 +32,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'minitest', '~> 5.3'
   spec.add_development_dependency "puma",     '~> 5.6'
   spec.add_development_dependency "wait_for_it",     '~> 0.1'
+  spec.add_development_dependency "timecop"
```
Comment (schneems) [inline barnes.gemspec:35]: Commenting a self-review

```
⏺ PR Review Summary: "Delay reporting 401 HTTP code for 90 seconds" (#62)

  Critical Issues (0 found)

  None. The implementation is correct and thread-safe.

  Important Issues (1 found)

  1. @first_401 is never cleared after recovery — lib/barnes/reporter.rb:91

  Once set, @first_401 persists for the lifetime of the Reporter instance. This means:
  - The 90-second grace period only works for the first burst of 401s
  - If the service recovers (200s) and then fails again later, the second burst will log immediately since Time.now - @first_401 will already exceed 90s

  This is arguably fine for the target scenario (dyno shutdown is one-way), but if you want the grace period to re-apply on each new burst, you could reset on success:

  when 200..299
    @first_401 = nil

  Suggestions (2 found)

  1. Boundary condition at exactly 90 seconds — The test checks 89s (silent) and 91s (prints), but not exactly 90s. The implementation uses > 90, so 90s stays silent. A test at now + 90 asserting
  silence would cheaply document this boundary, though it's operationally insignificant.

  2. Missing test for recovery-then-re-failure — A test showing: 401 → 200 (recovery) → 401 again would document the intended behavior of @first_401 never resetting. This serves as specification
  documentation more than bug prevention.

  Strengths

  - Clean dependency injection — The io: and debug: parameters make testing straightforward without global state mutation
  - Thread safety is correct — @first_401 is only accessed within the existing @mutex.synchronize block
  - Case ordering is correct — when 401 before when 400..499 means Ruby's first-match wins behavior works correctly
  - Good escape hatch — BARNES_DEBUG bypasses suppression for debugging
  - Appropriate scope — 401 is uniquely suited for this treatment among 4xx codes (other 4xx errors indicate permanent problems)
  - Tests cover the primary contract well — Timecop + StringIO is a solid combination for time-dependent output testing

  Recommended Action

  The PR is ready to merge as-is. The one "important" finding (no @first_401 reset) is a design choice that's correct for the stated use case — dyno shutdown is a terminal event. If you want to be
  defensive about long-lived processes with transient auth issues, adding @first_401 = nil in the 2xx branch is a one-line improvement but not required.
```

My reply:

- It is not meant to reset. If there's a transient 401 then I want it to be noisy. 
- Boundary condition at exactly 90 seconds test: This is overkill
- Missing test for recovery-then-re-failure: Also not really worth it. 

Basically, these are all questions about specifications and are answered by my reply here on this PR, and don't need to be explicitly added as test cases (IMO).

**Comment by schneems** [comment]
Fixing the race condition upstream (private repo).
